unblob: init at 25.1.8

vlaci opened the pull request:
Resolves: #217836

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

vlaci committed:
python3Packages.pyfatfs: init at 1.1.0

This is a dependency for Unblob (#217836)

vlaci committed:
unblob: init at 25.1.8

Resolves: #217836

vlaci commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,126 @@
+    ];
+
+    buildInputs = lib.optionals stdenv.hostPlatform.isDarwin [ libiconv ];
+    pythonImportsCheck = [ "unblob_native" ];
```
Moved it to its own package, so unblob's `flake.nix` can override it.

vlaci committed:
simg2img: 1.1.4 -> 1.1.5

To fix build on aarch64-darwin

This fixes broken tests of Unblob as well

vlaci commented:
Created a pair of this PR in Unblob's repo (
https://github.com/onekey-sec/unblob/pull/970). It runs CI for x86_64-linux, aarch64-linux, x86_64-darwin and aarch64-darwin exercising the code of this PR on those platforms.

SuperSandro2000 commented on pkgs/development/python-modules/pyfatfs/default.nix:
```diff
@@ -0,0 +1,49 @@
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
+
+  passthru.tests = {
+    pytest = callPackage ./tests.nix { };
```
Same here as above

SuperSandro2000 commented on pkgs/development/python-modules/unblob-native/default.nix:
```diff
@@ -0,0 +1,42 @@
+  ];
+
+  buildInputs = lib.optionals stdenv.hostPlatform.isDarwin [ libiconv ];
+  pythonImportsCheck = [ "unblob_native" ];
```
```suggestion

  pythonImportsCheck = [ "unblob_native" ];
```

vlaci commented on pkgs/by-name/un/unblob/tests.nix:
```diff
@@ -0,0 +1,28 @@
+    "--no-cov"
+    # `disabledTests` swallows the parameters between square brackets
+    # https://github.com/tytso/e2fsprogs/issues/152
+    "-k 'not test_all_handlers[filesystem.extfs]'"
```
See comment above, I could not use `disabledTests`, as it swallows everything in square brackets.

SuperSandro2000 replied:
oh yeah, normally we just eat the other tests.

vlaci replied:
I don't want to do this here, because these are the most important regression tests.

vlaci commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,103 @@
+    tests.pytest = callPackage ./tests.nix { };
+    updateScript = gitUpdater { };
+    # helpful to easily add these to a nix-shell environment
+    inherit runtimeDeps;
```
We need this for tests as well. Ideally it'd be the best to patch these paths in Unblob. This wrapping is one reason I didn't put this derivation in `pythonPackages`, as wrapping would be ineffective when used as a library.

vlaci commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,103 @@
+
+  doCheck = false;
+  passthru = {
+    tests.pytest = callPackage ./tests.nix { };
```
Inline as using `checkPhase` of the original derivation or inline as putting the test derivation inline in `passthru`?

I liked the approach that I can work on the derivation without the need of running tests.

SuperSandro2000 replied:
I don't know who wrote this but for pythonPackages we run python tests inside the normal build and only move them out if they cause an infinite recursion because otherwise they are not run when doing python updates to related packages and the failures might not be catched.

vlaci commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,103 @@
+    repo = "unblob";
+    rev = version;
+    hash = "sha256-r5UGoCOzYO/v7cAkbhOmeX7caspues/k/gokuzh1wsU=";
+    forceFetchGit = true;
```
No:
```
       error: function 'anonymous lambda' called with unexpected argument 'fetchLFS'
       at /nix/store/qhycnc3wyxszflkpkipnzma39zfpg3ih-source/pkgs/build-support/fetchurl/default.nix:52:1:
           51|
           52| {
             | ^
           53|   # URL to fetch.
```

SuperSandro2000 commented on pkgs/development/python-modules/pyfatfs/default.nix:
```diff
@@ -0,0 +1,49 @@
+
+  doCheck = false;
+
+  nativeBuildInputs = [
```
```suggestion
  build-system = [
```

SuperSandro2000 commented on pkgs/development/python-modules/pyfatfs/default.nix:
```diff
@@ -0,0 +1,49 @@
+    setuptools
+    setuptools-scm
+  ];
+  propagatedBuildInputs = [ fs ];
```
```suggestion

  dependencies = [ fs ];
```

SuperSandro2000 commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,107 @@
+  ];
+
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+        --replace-fail jefferson "# jefferson" \
+        --replace-fail ubi-reader "# ubi-reader"
```
```suggestion
    substituteInPlace pyproject.toml \
      --replace-fail jefferson "# jefferson" \
      --replace-fail ubi-reader "# ubi-reader"
```
we use 2 spaces even for shell code in nix files

SuperSandro2000 commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,107 @@
+  };
+
+  strictDeps = true;
+  buildInputs = with python3.pkgs; [ poetry-core ];
+
+  propagatedBuildInputs = with python3.pkgs; [
```
```suggestion
  build-system = with python3.pkgs; [ poetry-core ];

  dependencies = with python3.pkgs; [
```
this should be used with pyproject = true

vlaci commented on pkgs/by-name/un/unblob/tests.nix:
```diff
@@ -0,0 +1,28 @@
+    with python3.pkgs;
+    [
+      pytestCheckHook
+      pytest-cov
```
It won't work:

```
       > ImportError while loading conftest '/build/source/tests/conftest.py'.
       > tests/conftest.py:6: in <module>
       >     from unblob.testing import (  # noqa: F401 (module imported but unused)
       > unblob/testing.py:7: in <module>
       >     from pytest_cov.embed import cleanup_on_sigterm
       > E   ModuleNotFoundError: No module named 'pytest_cov.embed'
```

vlaci commented:
Sorry for the long absence, busy times. I hope, I could resolve everything satisfactorily.
I have rebased and re-tested on latest `master`, also updated `unblob` itself to the latest revision.

gepbird reviewed (changes requested):
Thanks! Mostly LGTM, just a few nits

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 345554`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 11 packages built:</summary>
  <ul>
    <li>python311Packages.pyfatfs</li>
    <li>python311Packages.pyfatfs.dist</li>
    <li>python311Packages.unblob-native</li>
    <li>python311Packages.unblob-native.dist</li>
    <li>python312Packages.pyfatfs</li>
    <li>python312Packages.pyfatfs.dist</li>
    <li>python312Packages.unblob-native</li>
    <li>python312Packages.unblob-native.dist</li>
    <li>simg2img</li>
    <li>unblob</li>
    <li>unblob.dist</li>
  </ul>
</details>

pluiedev commented on pkgs/development/python-modules/pyfatfs/default.nix:
```diff
@@ -0,0 +1,51 @@
+    description = "Python based FAT12/FAT16/FAT32 implementation with VFAT support";
+    homepage = "https://github.com/nathanhi/pyfatfs";
+    license = lib.licenses.mit;
+    platform = lib.platforms.all;
```
See [CI failure](https://github.com/NixOS/nixpkgs/actions/runs/12253963059/job/34183933678)
```suggestion
    platforms = lib.platforms.all;
```

vlaci replied:
yep,sorry for the typo

pluiedev commented on pkgs/development/python-modules/unblob-native/default.nix:
```diff
@@ -0,0 +1,45 @@
+    hash = "sha256-jpaBxKuQNfU0I3kCs67mM5dzGURSSHvqymhk43P7xXk";
+  };
+
+  cargoDeps = rustPlatform.importCargoLock {
+    lockFile = ./Cargo.lock;
+  };
```
Is this lockfile still necessary? We should avoid vendoring lockfiles if possible with mechanisms like `fetchCargoVendor`

vlaci replied:
changed to use `fetchCargoVendor`

pluiedev commented on pkgs/development/python-modules/unblob-native/default.nix:
```diff
@@ -0,0 +1,45 @@
+  passthru.updateScript = gitUpdater { rev-prefix = "v"; };
+
+  meta = {
+    description = "Performance sensitive parts of Unblob.";
```
```suggestion
    description = "Performance sensitive parts of Unblob";
```

pluiedev commented on pkgs/development/python-modules/pyfatfs/default.nix:
```diff
@@ -0,0 +1,51 @@
+  ];
+
+  env.SETUPTOOLS_SCM_PRETEND_VERSION = version;
+
```
updateScript?

gepbird commented on pkgs/development/python-modules/unblob-native/default.nix:
```diff
@@ -0,0 +1,46 @@
+
+  pythonImportsCheck = [ "unblob_native" ];
+
+  passthru.updateScript = gitUpdater { rev-prefix = "v"; };
```
This doesn't update the `cargoDeps` hash

vlaci replied:
I hope `nix-update-script` is the right solution :thinking:

gepbird replied:
Yes, it works now, thanks!

By the way to test the update script you can downgrade the version (doesn't matter if it doesn't exist), change the hashes (fails with empty has, so usually I just replace a few characters with zeros) then run `nix-shell maintainers/scripts/update.nix --argstr package python312Packages.unblob-native`. After the script finishes you should see the version and hashes changed back

vlaci replied:
Yep, did the same to test it :relieved:

gepbird reviewed (changes requested):
Thanks! Apart from the update script not working for `cargoDeps` and a few nits, I feel like this is done

(sorry about the `platform` typo)

pluiedev commented on pkgs/development/python-modules/unblob-native/default.nix:
```diff
@@ -0,0 +1,46 @@
+    cargoSetupHook
+  ];
+
+  buildInputs = lib.optionals stdenv.hostPlatform.isDarwin [ libiconv ];
```
Don't use `stdenv.is*`. See #341407

gepbird commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 345554`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 11 packages built:</summary>
  <ul>
    <li>python311Packages.pyfatfs</li>
    <li>python311Packages.pyfatfs.dist</li>
    <li>python311Packages.unblob-native</li>
    <li>python311Packages.unblob-native.dist</li>
    <li>python312Packages.pyfatfs</li>
    <li>python312Packages.pyfatfs.dist</li>
    <li>python312Packages.unblob-native</li>
    <li>python312Packages.unblob-native.dist</li>
    <li>simg2img</li>
    <li>unblob</li>
    <li>unblob.dist</li>
  </ul>
</details>

vlaci commented:
Sorry @gepbird, didn't want to re-request review from you as well :disappointed:

gepbird reviewed (approved):
No problem :)

pluiedev commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,128 @@
+    makeWrapper
+  ];
+
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace-fail jefferson "# jefferson" \
+      --replace-fail ubi-reader "# ubi-reader"
+  '';
```
These two seem like ordinary Python dependencies (i.e. not `build-system`) — `pythonRemoveDeps` should work here (if not, add a comment explaining why)
```suggestion
  pythonRemoveDeps = [
    "jefferson"
    "ubi-reader"
  ];
```

pluiedev reviewed (approved):
All looks good except for one last nitpick

SuperSandro2000 commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,129 @@
+    with python3.pkgs;
+    [
+      pytestCheckHook
+      pytest-cov
```
This could be pytest-cov-stub

gepbird replied:
Can you expand on that? Didn't work [before](https://github.com/NixOS/nixpkgs/pull/345554#discussion_r1795188207) nor does it work now `ModuleNotFoundError: No module named 'pytest_cov.embed'`

vlaci replied:
(background: tests need to setup coverage to work in subprocesses with signals, see https://pytest-cov.readthedocs.io/en/latest/subprocess-support.html#signal-handlers)

gepbird commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,129 @@
+
+  doInstallCheck = true;
+
+  nativeInstallCheckInputs = with python3.pkgs; [
+    versionCheckHook
+  ];
```
I'm not sure how this worked before
```suggestion
  nativeInstallCheckInputs = [
    versionCheckHook
  ];
```

vlaci replied:
I am not sure when and how this hook would run, but it doesn't seem to be executed.
If I replace this with this, the package still builds:

```nix
  nativeInstallCheckInputs = throw "nativeInstallCheckInputs";
```

vlaci replied:
Looks like, `builtPythonPackage` overwrites this: https://github.com/NixOS/nixpkgs/blob/5b37bd1fd84170665ea72db71ae33acb4312572d/pkgs/development/interpreters/python/mk-python-derivation.nix#L383

vlaci replied:
imo, we need to put it in `nativeCheckInputs`

pluiedev replied:
Yeah it should be in `nativeCheckInputs` for Python packages but `nativeInstallCheckInputs` anywhere else

gepbird commented on pkgs/by-name/un/unblob/package.nix:
```diff
@@ -0,0 +1,129 @@
+  simg2img,
+  ubi_reader,
+  unar,
+  zstd,
```
```suggestion
  zstd,
  versionCheckHook,
```

vlaci commented:
What can I do, to push this in for the holidays? :)

@SuperSandro2000 you are the last one with outstanding comments.

FliegendeWurst committed:
simg2img: fix GCC 14 build

vlaci commented:
> It seems unblob is not compatible with current library versions
> 
> ```
> Checking runtime dependencies for unblob-24.12.4-py3-none-any.whl
>   - cryptography<44.0,>=41.0 not satisfied by version 44.0.0
>   - lief<0.16.0,>=0.15.1 not satisfied by version 0.16.1
> ```
> 
> You may need to add these to `pythonRelaxDeps` (check afterwards that the program still works).

Lief needs code changes as well. These are fixed on the `main` branch as well. Will release a new version soon(tm)

vlaci commented:
Updated to master, fixed breakages.

FliegendeWurst commented:
Oh, I already did #372073 for simg2img

FliegendeWurst commented on pkgs/by-name/si/simg2img/package.nix:
```diff
@@ -13,13 +19,42 @@ stdenv.mkDerivation rec {
+      url = "https://github.com/anestisb/android-simg2img/commit/931df9dd83e7feea11197402c5b4e7ad489f4abf.patch";
+      hash = "sha256-+vHdx+nwaeaLf7a/hZSAVECc/Q5nQ1G2i0eTOVHb8Ks=";
+    })
+    (fetchpatch {
+      name = "PR-41-fix-calloc-argument-order-for-gcc-14-1.patch";
+      url = "https://github.com/anestisb/android-simg2img/commit/800987799591acd76d7d33853f665873aa5d63be.patch";
+      hash = "sha256-5dEmZY6/MiKbxQu1amR9F8pAFhX4T+10RNMTxKaVbc0=";
+    })
+    (fetchpatch {
+      name = "PR-41-fix-calloc-argument-order-for-gcc-14-2.patch";
+      url = "https://github.com/anestisb/android-simg2img/commit/bfcfd2cc3e249c0a1e46d466c57b1dc5ee352291.patch";
+      hash = "sha256-Ot5Uij/Je1gS0e9f5zP5+GPAIRl94vew87VMjd0J3ic=";
+    })
+    (fetchpatch {
+      name = "PR-41-fix-calloc-argument-order-for-gcc-14-3.patch";
+      url = "https://github.com/anestisb/android-simg2img/commit/4fcde3f0cebbc9466f1e73de04d82bc548dd1414.patch";
+      hash = "sha256-h9fr5H6hfi9SqKK9T7dvHiLK3wRKnsxkczwS71Ubh/g=";
+    })
```
These are not merged yet. You can copy the postPatch from my PR instead

vlaci replied:
When https://github.com/NixOS/nixpkgs/pull/372073 goes in, I'll drop it.

I feel great disturbance in the force, when for some review comments I get that I should use patches when they are available instead of `postPatch` commands. (sigh)

FliegendeWurst replied:
That is indeed the correct action, if they are merged in the upstream repository.
Right now I can force push to my branch, and all the patches would disappear.

vlaci replied:
Never mind, cherry-picked that patch instead.

FliegendeWurst commented on pkgs/by-name/si/simg2img/package.nix:
```diff
@@ -13,13 +19,42 @@ stdenv.mkDerivation rec {
+
+  patches = [
+    (fetchpatch {
+      name = "PR-38-Fix-for-Apple-Silicon.patch";
```
This one is included in 1.1.5, maybe we should just update to that version?

vlaci replied:
Nice!

vlaci committed:
simg2img: apply nixfmt

vlaci commented:
> Oh, I already did #372073 for simg2img

:see_no_evil: didn't know that

lafrenierejm commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 345554`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 11 packages built:</summary>
  <ul>
    <li>python312Packages.pyfatfs</li>
    <li>python312Packages.pyfatfs.dist</li>
    <li>python312Packages.unblob-native</li>
    <li>python312Packages.unblob-native.dist</li>
    <li>python313Packages.pyfatfs</li>
    <li>python313Packages.pyfatfs.dist</li>
    <li>python313Packages.unblob-native</li>
    <li>python313Packages.unblob-native.dist</li>
    <li>simg2img</li>
    <li>unblob</li>
    <li>unblob.dist</li>
  </ul>
</details>
